jdbc-driver/src/main/java/com/google/cloud/dataproc/jdbc/DataprocInfo.java

```diff
@@ -0,0 +1,188 @@
+        String host =
+                params.clusterName() == null
+                        ? findClusterInPool(formatClusterFilterString())
+                        : getClusterByName();
+        return host + "-m";
+    }
+
+    /**
+     * Supports finding & verifying the cluster indicated by name.
+     *
+     * @return the cluster of the given name
+     * @throws InvalidURLException
+     */
+    public String getClusterByName() throws SQLException {
+        try {
+            Cluster cluster =
+                    clusterControllerClient.getCluster(
+                            params.projectId(), params.region(), params.clusterName());
+            return cluster.getClusterName();
+        } catch (ApiException e) {
+            if (e.getStatusCode().getCode().equals(StatusCode.Code.NOT_FOUND)) {
+                throw new InvalidURLException(
+                        String.format(
+                                "Unable to retrieve cluster information for %s/%s/%s.\n",
+                                params.projectId(), params.region(), params.clusterName()),
+                        e);
+            } else {
+                throw new SQLException(e);
+            }
+        }
+    }
+
+    /**
+     * Formats the user passed in clusterPoolLabel value to the filter format that Dataproc API
+     * accpets: field = value [AND [field = value]] ...
+     *
+     * @return formatted filter string accepted by Dataproc Client API
+     */
+    public String formatClusterFilterString() throws InvalidURLException {
+        Map<String, List<String>> labels = new LinkedHashMap<>();
+        // `ACTIVE` contains the `CREATING`, `UPDATING`, and `RUNNING` states
+        labels.put("status.state", ImmutableList.of("ACTIVE"));
+
+        // clusterPoolLabel can be null, but filter is not empty string since
+        // it at least has "status.state = ACTIVE"
+        if (params.clusterPoolLabel() != null) {
+            for (String label : params.clusterPoolLabel().split(":")) {
+                checkUrl(
+                        !label.isEmpty() && label.contains("="),
+                        "'%s' Invalid clusterPoolLabel.\n"
+                                + "Example format: clusterPoolLabel=field1=value1:field2=value2'",
+                        label);
+                String field = label.split("=")[0];
+                String value = label.split("=")[1];
+
+                if (field.equals("status.state")) {
+                    throw new InvalidURLException(
+                            "Please do not provide cluster status as label, "
+                                    + "since DataprocDriver will use cluster with status.state = ACTIVE.");
+                }
+
+                // filed can be label.[key] or clusterName
+                labels
+                        .computeIfAbsent(
+                                field.equals("clusterName") ? field : "labels." + field, k -> new ArrayList<>())
```

Comment:
May you use `StringBuilder` instead of `ArrayList` as a value, so you can create formatted filter string as you go, instead of having a separate `flatMap` step below?

Reply:
Changed, check it out in the new commit. But I'm not sure if the way I "create formatted filter string as you go" is elegant. Lmk what you think or any suggestion to make it better, thanks~

Reply:
Nvm! Labels should not contain duplicates and I'll change it.